CSP/rules.py

```diff
@@ -0,0 +1,92 @@
+                                  [variables[0], variables[1]])
+        elif len(variables) == 3:
+            problem.addConstraint(lambda x, y, z, r=required_pitches: set(r).issubset({x % 12, y % 12, z % 12}),
+                                  [variables[0], variables[1], variables[2]])
+        elif len(variables) == 4:
+            problem.addConstraint(lambda w, x, y, z, r=required_pitches: set(r).issubset({w % 12, x % 12, y % 12, z % 12}),
+                                  [variables[0], variables[1], variables[2], variables[3]])
+
+
+def parallels(voices, problem, names, interval):
+    for i in range(0, len(voices[0]) - 1):  # Doesn't apply to last pitches (final cadence)
+        voice_index = 0
+        for voice in voices:
+            pitch = voice[i]
+            pitch_next = voice[i + 1]
+            if pitch == '??':
+                pitch_variable = names[voice_index] + str(i + 1)
+                for j in [x for x in range(0, 4) if x != voice_index]:
+                    other_pitch = voices[j][i]
+                    other_pitch_next = voices[j][i + 1]
+                    if other_pitch == '??':
+                        other_pitch_variable = names[j] + str(i + 1)
+                        if pitch_next == '??':
+                            pitch_next_variable = names[voice_index] + str(i + 2)
+                            if other_pitch_next == '??':
+                                other_pitch_next_variable = names[j] + str(i + 2)
+                                problem.addConstraint(
+                                    lambda x1, x2, y1, y2:
+                                    not (abs(x1 - y1) % 12 == interval and interval == abs(x2 - y2) % 12 and x1 != x2),
+                                    [pitch_variable, pitch_next_variable, other_pitch_variable, other_pitch_next_variable])
+                    else:
+                        if pitch_next == '??':
+                            pitch_next_variable = names[voice_index] + str(i + 2)
+                            if other_pitch_next == '??':
+                                other_pitch_next_variable = names[j] + str(i + 2)
+                                problem.addConstraint(
+                                    lambda x1, x2, y2, y1=other_pitch:
+                                    not (abs(x1 - y1) % 12 == interval and interval == abs(x2 - y2) % 12 and x1 != x2),
+                                    [pitch_variable, pitch_next_variable, other_pitch_next_variable])
+                            else:
+                                problem.addConstraint(
+                                    lambda x1, x2, y1=other_pitch, y2=other_pitch_next:
+                                    not (abs(x1 - y1) % 12 == interval and interval == abs(x2 - y2) % 12 and x1 != x2),
+                                    [pitch_variable, pitch_next_variable])
+                # TODO: Implement extended cases
```

Comment:
This is todo for possible later improvement of project, right?

Reply:
Yup